compiler/rustc_symbol_mangling/src/errors.rs

```diff
@@ -0,0 +1,36 @@
+//! Errors emitted by symbol_mangling.
+
+use rustc_macros::SessionDiagnostic;
+use rustc_span::Span;
+
+#[derive(SessionDiagnostic)]
+#[error(symbol_mangling::invalid_symbol_name)]
+pub struct InvalidSymbolName {
+    #[primary_span]
+    pub span: Span,
+    pub mangled_formatted: String,
+}
+
+#[derive(SessionDiagnostic)]
+#[error(symbol_mangling::invalid_trait_item)]
+pub struct InvalidTraitItem {
+    #[primary_span]
+    pub span: Span,
+    pub demangling_formatted: String,
+}
+
+#[derive(SessionDiagnostic)]
+#[error(symbol_mangling::alt_invalid_trait_item)]
+pub struct AltInvalidTraitItem {
+    #[primary_span]
+    pub span: Span,
+    pub alt_demangling_formatted: String,
+}
```

Comment:
Quoting from the [commit message](https://github.com/rust-lang/rust/pull/100831/commits/020c6430b68c30c5c70479e7c06d82f97e2ef049):

> Thought of doing this by having a struct and an enum with Default and Alt cases, but not sure if we wanted to have the text in code instead of having “demangling()” and “demangling-alt()” in the ftl file.
> 
> Don’t like the current way of having structs representing the same-ish and using long names to distinguish their expectations, instead of putting this in an enum and handling the different cases inside the type.
> 
> I am fine with whichever option the team prefers; also understand having them as separate structs keeps it simple.

Reply:
I think this is okay for this crate, it's certainly something that we've run into before in other crates there where it is more of an issue. It's a big job, but if you wanted to extend `SessionDiagnostic` to support enums (in a similar way to how `SessionSubdiagnostic` does - which might make it possible to merge the code behind those somewhat, maybe, a little) then I'd like to see that :)